gpytorch/lazy/triangular_lazy_tensor.py

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python3
+
+import torch
+
+from ..utils.broadcasting import _mul_broadcast_shape
+from ..utils.memoize import cached
+from .added_diag_lazy_tensor import AddedDiagLazyTensor
+from .lazy_tensor import LazyTensor
+from .non_lazy_tensor import NonLazyTensor
+
+
+class TriangularLazyTensor(LazyTensor):
+    def __init__(self, tensor, upper=False):
+        """
+        Triangular lazy tensor. Supports arbitrary batch sizes.
+
+        Args:
+            :attr:`tensor` (Tensor or LazyTensor):
+                A `b1 x ... x bk x n x n` Tensor, representing a `b1 x ... x bk`-sized batch
+                of `n x n` triangular matrices.
+            :attr:`upper` (bool):
+                If True, the tensor is upper-triangular, otherwise lower-triangular.
+        """
+        super().__init__(tensor)
+        if torch.is_tensor(tensor):
+            tensor = NonLazyTensor(tensor)
+        self._upper = upper
+        self._tensor = tensor
+
+    def __add__(self, other):
+        if isinstance(other, TriangularLazyTensor) and not self._upper ^ other.upper:
+            return self.__class__(self._tensor + other._tensor, upper=self._upper)
+        return self._tensor + other
+
+    @cached(name="cholesky")
+    def _cholesky(self):
+        return self.__class__(self._tensor._cholesky(), upper=self._upper)
```

Comment:
In general - we could maybe add a subclass for tensors that we do not expect to be PSD square matrices.

Reply:
I guess what I really should do is have `DiagLazyTensor` subclass `TriangularLazyTensor`, then the cholesky issue is handled correctly.

Reply:
re the non-PSD square matrices: This would make sense but also create a bunch of complications. Like how to handle things like `SumLazyTensor` - would we mark it PSD or non-PSD depending on whether all constituent tensors are psd (sufficient)?